File: java/ABC3Thread.java
public class ABC3Thread{
    private static class Cond{
        private int mCurrent;
        private byte [] mutex = new byte[0];
        private int mMax;
        private int getCurrent(){
            synchronized(mutex){
                return mCurrent;
            }
        }
        private void next(){
            synchronized(mutex){
                mCurrent++;
                if (mCurrent == mMax)
                    mCurrent = 0;
            }
        }
        public Cond(int m){
            mMax = m;
        }
    }
    private static class PrintThread implements Runnable{
        private String mInfo;
        private int mId;
        private Cond mCond;
        public PrintThread(String info, int id, Cond c){
            mInfo = info;
            mId = id;
            mCond = c;
        }
        @Override
        public void run(){
            for (int i = 0; i < 10 ; i++) {
                synchronized(mCond){
                    try{
                        while(mCond.getCurrent() != mId){
                            mCond.wait();
                        }
                    }catch(Exception e){
                        e.printStackTrace();
                        return;
                    }
                }
                System.out.println(mInfo);
                synchronized(mCond){
                    mCond.next();
                    mCond.notifyAll();
                }
            }
        }
    }

    public static void main(String [] argv){
        Cond cond = new Cond(3);
        PrintThread A = new PrintThread("A", 0, cond);
        PrintThread B = new PrintThread("B", 1, cond);
        PrintThread C = new PrintThread("C", 2, cond);
        new Thread(A).start();
        new Thread(B).start();
        new Thread(C).start();
        synchronized(cond){
            cond.notifyAll();
        }
    }
}

File: thread/ABC3Thread.py
#!/usr/bin/python3
import threading

class ID:
    def __init__(self, lock, maxid):
        self.lock = lock
        self.id = 1
        self.maxid = maxid

    def getCurrent(self):
        self.lock.acquire()
        tid = self.id
        self.lock.release()
        return tid

    def next(self):
        self.lock.acquire(1)
        self.id = self.id + 1
        if self.id > self.maxid:
            self.id = 1
        self.lock.release()

class PrintThread(threading.Thread):
    def __init__(self, myid, ID, info, cond):
        threading.Thread.__init__(self)
        self.myid = myid
        self.ID = ID
        self.info = info
        self.cond = cond

    def run(self):
        i = 0
        while i < 10: 
            while not self.ID.getCurrent() == self.myid:
                cond.acquire()
                cond.wait()
                cond.release()
            print('%s: %d' % (self.info, i))
            self.ID.next()
            cond.acquire()
            cond.notifyAll()
            cond.release()
            i = i + 1


if __name__ == '__main__':
    ID = ID(threading.Lock(), 3)
    cond = threading.Condition()
    A = PrintThread(1, ID, 'A', cond)
    B = PrintThread(2, ID, 'B', cond)
    C = PrintThread(3, ID, 'C', cond)
    A.start()
    B.start()
    C.start()

